web3/contract.py

```diff
@@ -1668,9 +2326,27 @@ def find_functions_by_identifier(
     ]
 
 
+def async_find_functions_by_identifier(
+    contract_abi: ABI, w3: 'Web3', address: ChecksumAddress, callable_check: Callable[..., Any]
+) -> List[AsyncContractFunction]:
+    fns_abi = filter_by_type('function', contract_abi)
+    return [
+        AsyncContractFunction.factory(
```

Comment:
yea, I really went back and forth here on how much to make the BaseContract aware of the types AsyncContractFunction and ContractFunction since we are using typing. But if we go this route it will definitely clean up the code some.